---
title: "Bioinformatic analyses of plasmid-borne RsmA"
author: "jpjh"
date: "November 2021"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(ggplot2)
library(knitr)

source("./func/theme_pub.r")
theme_github <- function(x){
  theme_pub(base_size=12)
}

theme_set(theme_github())
```

### 1.1 Identification of RsmA homologues from plasmids in databases

Questions:

- What plasmids are RsmA homologues on?
- What are the sequences of these genes?
- What kinds of plasmids are these sequences on?

COMPASS plasmid database was annotated using [PROKKA](https://github.com/tseemann/prokka), with default settings: PROKKA uses UniProtKB (SwissProt) databases to annotate.

Link to annotation in `./ref/COMPASS_prokka.gff` and `./ref/COMPASS_prokka.faa`.

Identify genes annotated as 'CsrA' or 'Carbon storage regulator'

```{bash, eval=FALSE}
grep "[Cc]srA" ./ref/COMPASS_prokka.gff | wc -l # 106
grep "[Cc]arbon storage regulator" ./ref/COMPASS_prokka.gff | wc -l # 11
```

Note: searching for '[Rr]smA' gives 14 hits, 'all Ribosomal RNA small subunit methyltransferase A', which is irrelevant.

Note: no additional matches when looking for "[Cc]arbon storage".

Note: all hits with "[Cc]arbon storage regulator" also matched "[Cc]srA".

This gives us our list of CsrA-like genes in COMPASS. Need to match these up with the overall COMPASS metadata to get a table with plasmid details and CsrA sequence. 

```{bash, eval=FALSE}
grep "[Cc]srA\|[Cc]arbon storage regulator" ./ref/COMPASS_prokka.faa \
  | sed 's/>//g' > ./working_1/COMPASS_CsrA_matches.txt
  
seqtk subseq ./ref/COMPASS_prokka.faa \
  ./working_1/COMPASS_CsrA_matches.txt \
  | awk '{print $1}' \
  | tr '\n' '\t' | tr '>' '\n' \
  > ./working_1/COMPASS_CsrA_matches.tsv
  
seqtk subseq ./ref/COMPASS_prokka.ffn \
  ./working_1/COMPASS_CsrA_matches.txt \
  | awk '{print $1}' \
  | tr '\n' '\t' | tr '>' '\n' \
  > ./working_1/COMPASS_CsrA_matches_na.tsv  
  
grep "[Cc]srA\|[Cc]arbon storage regulator" ./ref/COMPASS_prokka.gff \
  > ./working_1/COMPASS_prokka_CsrA_matches.gff
```

Load into R and join with plasmid details. 

Note: using the full table from COMPASS to join, which probably results in some duplicates.

```{r, eval=FALSE}
# Note: eval is set to FALSE as a workaround to deal with the fact 
#    that Rmarkdown doesn't work with symbolic links. 
# This code was run in advance to prepare the local .rds files for analysis

source("./func/gff_functions.R")

comp <- read.csv("./ref/COMPASS.csv", row.names = NULL, sep=";")
dups <- read.csv("./ref/COMPASS_duplicates_by_accession.txt", header=FALSE)
gff <- gffRead("./working_1/COMPASS_prokka_CsrA_matches.gff")

write_rds(file="./working_1/comp.rds", comp)
write_rds(file="./working_1/dups.rds", dups)
write_rds(file="./working_1/gff.rds", gff)
```

Make some diagnostic plots. Note: `comp` is the full COMPASS database, `compass` is with duplicates removed. 

```{r}
source("./func/gff_functions.R")

csra_seqs <- read.table("./working_1/COMPASS_CsrA_matches.tsv",
                        col.names=c("ID","seq"))

comp <- read_rds("./working_1/comp.rds")
dups <- read_rds("./working_1/dups.rds")
gff <- read_rds("./working_1/gff.rds")

compass <- comp %>% filter(!(Accession %in% dups$V1))

csra_gff <- data.frame(Accession = gff$seqname,
                       start = gff$start,
                       end = gff$end,
                       strand = gff$strand,
                       ID = getAttributeField(gff$attributes, "ID"))

# Note: using the full table from COMPASS to join, which probably results in some duplicates.

plasmid_csra <- csra_gff %>% left_join(csra_seqs, by="ID") %>%
  left_join(comp, by="Accession")

plasmid_csra %>% filter(duplicated(seq)) %>% nrow() # 55 sequences are identical to others in the database at the AA level

# are these duplications within Genus, or are there duplicates in different Genera?

plasmid_csra %>% group_by(seq, Genus) %>% count() %>% filter(duplicated(seq)) %>% nrow() 

# No duplicates across genus, so all duplicates are duplicated within Genus.

plasmid_csra <- plasmid_csra %>% mutate(Duplicated = factor(ifelse(duplicated(seq), "Y", "N"),
                                                            levels=c("Y","N")))

write.table(unique(plasmid_csra$Accession), 
            file="./working_1/plasmid_csra_accessions.txt", col.names = FALSE,
            quote=FALSE, row.names = FALSE)
```

The NA is a Xanthomondales bacterium, [*Ahniella afigens*](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=2021234) (Rhodanobacteriaceae). Manually annotate this line. 

```{r}
plasmid_csra %>% filter(is.na(Genus))

compass[compass$Accession=="NZ_CP027861.1",
        c("Family","Genus")] <- c("Rhodanobacteraceae","Ahniella")

plasmid_csra[plasmid_csra$Accession=="NZ_CP027861.1",
             c("Family","Genus")] <- c("Rhodanobacteraceae","Ahniella")
```

Get overall statistics for distribution.

```{r}
nrow(plasmid_csra) # 106 csrA genes total

plasmid_csra %>% filter(!duplicated(Accession)) %>% nrow() # across 98 plasmids

plasmid_csra %>% group_by(Class, Accession) %>% summarise(n = n()) %>% 
  summarise(n = n()) %>% kable() #92/98 from Gammaproteobacteria
```

Plot CsrA homologues across Genera.

```{r}
ggplot(data=plasmid_csra, aes(x=Genus)) + geom_bar() + theme_pub() +
  theme(axis.text.x = element_text(face="italic", angle=45, hjust=1))

(plot_bar_Class <- ggplot(data=plasmid_csra, aes(x=Genus, fill=Class, alpha=Duplicated)) + geom_bar() + 
  scale_fill_brewer(type="qual", palette=6, name="") + 
  labs(y = "# plasmid-borne CsrA/RsmA homologues") + 
  scale_alpha_manual(values=c(0.5,1)) + theme_pub() +
  theme(axis.text.x = element_text(face="italic", angle=45, hjust=1), legend.position="right"))

png(filename="./figs/1_1_bar_by_genus.png", height=3, width=4.5, units="in", res=300)
plot_bar_Class + theme_pub() +
  theme(axis.text.x = element_text(face="italic", angle=45, hjust=1), legend.position="right")
dev.off()
```

Shows almost all (>93%) come from Gammaproteobacteria. Within Gammaproteobacteria, about 40% have at least one duplicate at the amino-acid level in the database.

How does this compare with plasmid size and other features? Plot CsrA-containing plasmids onto a density plot for each Family. For this, remove the duplicates, because they can confuse the density plot, and remove all Families with <20 plasmids.

```{r}
compass <- compass %>% mutate(encodes_csra = ifelse(Accession %in% plasmid_csra$Accession, "Y", "N"))

compass %>% filter(encodes_csra == "Y") %>% arrange(Size..bp.) %>% select(Description, Size..bp.) %>%
  head() %>% kable()

csra_families <- plasmid_csra$Family %>% unique()

compass_subset_fam <- filter(compass, Family %in% csra_families)

(counts_family <- compass_subset_fam %>% group_by(Family, encodes_csra) %>% 
    summarise(total = n()) %>% pivot_wider(names_from=encodes_csra, values_from=total, values_fill=0) %>%
  mutate(totals = N + Y)) %>% kable()

counts_family_20 <- filter(counts_family, totals>20)

compass_subset_fam_20 <- filter(compass_subset_fam, 
                             Family %in% counts_family_20$Family)

(plot_joy_family <- ggplot(compass_subset_fam_20, aes(x=log10(Size..bp.))) +
  geom_density(size=0, aes(fill=Class), alpha=0.8) +
  scale_fill_brewer(type="qual", palette=6, name="") + 
    # geom_density(data=subset(compass_subset_fam_20, encodes_csra=="Y"), size=0.5, colour="black") +
  geom_point(data=subset(compass_subset_fam_20, encodes_csra=="Y"), y=0.1,
             shape=17, alpha=0.6, colour="black") +
  expand_limits(y=c(0,0.5)) +
  facet_grid(Family~., scales="free_y") + 
  geom_text(data=counts_family_20, x=3, y=0, hjust=0, vjust=-0.5, aes(label=paste("n = ", Y, " / ", totals, sep="")), size=2) + 
  labs(x=expression(paste("log"[10], " plasmid size in bp")), y="") + 
  theme(axis.line.y = element_blank(), axis.text.y = element_blank(),
        panel.border = element_blank(), axis.ticks.y = element_blank(), legend.position="bottom"))

png(filename="./figs/1_2_distribution_family.png", height=3, width=4.5, units="in", res=300)
plot_joy_family + theme_pub() +
  theme(axis.line.y = element_blank(), axis.text.y = element_blank(),
        panel.border = element_blank(), axis.ticks.y = element_blank(), legend.position="bottom")
dev.off()
```

Are there any obvious correlations between CsrA and mobility?

```{r}
(counts_family_mobility <- compass_subset_fam_20 %>% group_by(Family, Predicted.Mobility, encodes_csra) %>% 
   summarise(total = n()) %>% pivot_wider(names_from=encodes_csra, values_from=total, values_fill=0) %>%
   mutate(totals = N + Y, label = paste(Y, totals, sep="/"))) %>% kable()
```

No clear patterns, and most Families have only a couple of CsrA-encoding plasmids. Test statistically using Fisher's Exact Test.

```{r}
run_fisher_on_mobility_families <- function(x){
  test = x %>%
    column_to_rownames("Predicted.Mobility") %>% 
    select(N, Y) %>% as.matrix() %>%
    fisher.test(simulate.p.value=TRUE)
  return(test$p.value)
}

counts_family_mobility %>% group_by(Family) %>%
  summarise(p.value = run_fisher_on_mobility_families(cur_data())) %>%
  mutate(p.adj = p.adjust(p.value, method="bonferroni")) %>% kable()
```

Significant effect for Legionellaceae (more CsrA on conjugative plasmids, p_adj = 0.02), but this could be due to repeated sampling of outbreak strains. Plot data.

```{r}
(mobility_csra <- ggplot(data=compass_subset_fam_20, aes(fill=Predicted.Mobility, x=Predicted.Mobility)) + 
    geom_bar(aes(alpha=encodes_csra)) + 
    scale_alpha_manual(values=c(0.5,1), name="Encodes CsrA?") +
    scale_fill_manual(values=c("forestgreen","darkorange","red"), name="Predicted mobility") +
    labs(x="") +
    # geom_text(data=counts_family_mobility, aes(y=totals, label=label), size=2) +
    facet_wrap(~Family, scales="free_y", ncol=4) + 
    theme(legend.position="bottom", axis.text.x=element_text(angle=45, hjust=1)))

(mobtype_csra <- ggplot(data=compass_subset_fam_20, aes(alpha=encodes_csra, fill=Predicted.Mobility, x=MOB.types)) + 
  scale_alpha_manual(values=c(0.5,1), name="Encodes CsrA?") +
  scale_fill_manual(values=c("forestgreen","darkorange","red"), name="Predicted mobility") +
  labs(x="") +
  facet_wrap(~Family, scales="free_y", ncol=4) + 
  geom_bar() + theme(legend.position="bottom", axis.text.x=element_text(angle=45, hjust=1)))

(mpftype_csra <- ggplot(data=compass_subset_fam_20, aes(alpha=encodes_csra, fill=Predicted.Mobility, x=MPF.types)) + 
  scale_alpha_manual(values=c(0.5,1), name="Encodes CsrA?") +
  scale_fill_manual(values=c("forestgreen","darkorange","red"), name="Predicted mobility") +
  labs(x="") +
  facet_wrap(~Family, scales="free_y", ncol=4) + 
  geom_bar() + theme(legend.position="bottom", axis.text.x=element_text(angle=45, hjust=1)))

png(filename="./figs/1_3_mobility_csra.png", height=3, width=4.5, units="in", res=300)
mobility_csra + theme_pub() +
  theme(legend.position="bottom", axis.text.x=element_text(angle=45, hjust=1))
dev.off()
```

Plot numbers of CsrA plasmids for all Families with >20 plasmids in COMPASS.

```{r, warning=FALSE}
(compass_totals <- compass %>% group_by(Family) %>% count() %>% arrange(-n))

counts_compass_family_20 <- filter(compass_totals, n>20)

# Fix 'Class' of the Cyanobacteriaceae based on http://www.marinespecies.org/

(compass_all_family_csra <- compass %>% 
    mutate(Class = ifelse(Family %in% c("Cyanothecaceae","Nostocaceae","Synechococcaceae"), "Cyanophyceae", Class)) %>%
    filter(Family %in% counts_compass_family_20$Family) %>%
  ggplot(aes(alpha=encodes_csra, fill=Class, x=fct_reorder(Family, Class, .fun = unique))) +
  scale_alpha_manual(values=c(0.5,1), name="Encodes CsrA?") + 
  labs(x="", y="# plasmids in COMPASS") +
  geom_bar() + theme(legend.position="bottom", axis.text.x=element_text(angle=45, hjust=1)))
  
library(patchwork)

png(filename="./figs/1_4_all_family.png", height=6, width=7.2, units="in", res=300)
(compass_all_family_csra + theme_pub() + 
    theme(axis.text.x=element_blank(), legend.position="none",
           plot.margin = margin(0,0,0,0.2, "in"))) /
  (compass_all_family_csra + theme_pub() + 
     theme(legend.position="bottom", axis.text.x=element_text(angle=45, hjust=1),
           plot.margin = margin(0,0,0,0.2, "in")) + 
     coord_cartesian(ylim=c(0,200)))
dev.off()
```

Interesting that of the ~3.6k Enterobacteriaceae plasmids, none have a CsrA homologue, whereas approx. 50% and 25% of Legionellaceae and Pseudomonadaceae plasmids do. Basic analysis:

```{r}
(counts_family_all_20 <- compass %>%
   filter(Family %in% counts_compass_family_20$Family & !(is.na(Family))) %>%
   group_by(Family, encodes_csra) %>% 
   summarise(total = n()) %>% pivot_wider(names_from=encodes_csra, values_from=total, values_fill=0) %>%
  mutate(totals = N + Y, ratio = Y/totals)) %>% kable()

counts_family_all_20 %>% column_to_rownames("Family") %>% 
  select(N, Y) %>% as.matrix() %>%
  fisher.test(simulate.p.value=TRUE)
```

Significant association of CsrA-containing plasmids with particular Families, though note that this doesn't capture the variation in sampling effort and potential for phylogenetic correlation in the database.

### 1.2 Analysis of diversity of plasmids encoding CsrA

Calculated MASH distances for all plasmids in COMPASS.

```{bash, eval=FALSE}
mash sketch ./ref/COMPASS.fasta -i -S 42 -s 1000 -k 21 -p 4 -o ../ref/COMPASS.msh
mash triangle ./ref/COMPASS.msh -i -k 21 -p 4 > ../ref/COMPASS_mash.dst
```

```{r, eval=FALSE}
mash_dist <- read.table("./ref/COMPASS_mash.dst", skip=1,
                        fill = TRUE, col.names=c("V0",paste("V", 1:12084, sep="")))
mash_dist <- column_to_rownames(mash_dist, "V0")

write.table(mash_dist, file = "./ref/COMPASS_mash.mat",
            row.names=FALSE, col.names=FALSE, quote=FALSE)

distmat <- as.dist(mash_dist, upper=TRUE, diag=TRUE)
sqmat <- as.matrix(distmat)

write.table(sqmat, file = "./ref/COMPASS_mash.sqmat",
            quote=FALSE)
```

Cluster with UMAP.

Note on how [Galeta et al.](https://pubmed.ncbi.nlm.nih.gov/30380090/) made their plot:

>First, pairwise distances between the sequences were computed from the created Mash sketches. Then, UMAP was applied to the distance matrix using parameters n_neighbors=50, n_components=2, init=‘random’, metric=‘precomputed’.

Activated `umap_env` in Conda, and then activated Python3.

```{python, eval=FALSE}
import numpy as np
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

import umap

reducer = umap.UMAP(n_neighbors=50, n_components=2, init='random', metric='precomputed')

distmat = pd.read_csv('./ref/COMPASS_mash.sqmat', sep=' ', header=0, index_col=0)

embedding = reducer.fit_transform(distmat)

output=pd.DataFrame(data={'Accession': distmat.index.values, 'x':embedding[:,0], 'y':embedding[:,1]})

output.to_csv("./ref/COMPASS_mash_UMAP.csv")
```

Load into R and plot.

```{r, eval=FALSE}
# again, load this item before knitting the Rmarkdown as the csv is a symbolic link

compass_mash_UMAP_ <- read.csv("./ref/COMPASS_mash_UMAP.csv",
                              row.names=1) 
write_rds(file="./working_1/compass_mash_umap.rds", compass_mash_UMAP_)
```

```{r}
compass_mash_UMAP_ <- read_rds("./working_1/compass_mash_umap.rds")

families_to_highlight <- c(unique(compass_subset_fam$Family))

compass_mash_UMAP <- compass_mash_UMAP_ %>%
  left_join(select(compass, Accession, Order, Family, Plasmid, Size..bp.)) %>% 
  arrange(Family) %>%
  mutate(encodes_csra = ifelse(Accession %in% plasmid_csra$Accession, "Y", "N"),
         fam_highlight = factor(ifelse(Family %in% families_to_highlight,
                                Family, "Other"), levels=c(families_to_highlight,"Other")))

library(scales)
library(ggrepel)

(family_umap <- ggplot(data=compass_mash_UMAP, aes(x=x, y=y, colour=fam_highlight, alpha=fam_highlight)) + 
  geom_point(shape=16) + 
  scale_alpha_manual(values=c(rep(0.8, length(families_to_highlight)), 0.2), name="") +
  scale_colour_manual(values=c(hue_pal()(length(families_to_highlight)), "grey50"), name="") +
  geom_point(data=subset(compass_mash_UMAP, encodes_csra=="Y"), 
             aes(fill=fam_highlight),
             shape=21, colour="black", show.legend = FALSE) +
  geom_text_repel(data=subset(compass_mash_UMAP, 
                              Accession %in% c("AM235768.1","NC_022344.1","NC_006365.1",
                                               "NZ_CP011850.1","NC_007507.1","NZ_CP019045.1",
                                               "NZ_CP014843.1")),
                  aes(label=Plasmid), colour="black", box.padding = 1, max.overlaps = Inf, size=2.5, show.legend=FALSE) +
  theme(legend.position="bottom"))

png(filename="./figs/1_5_family_umap.png", height=4.5, width=4.5, units="in", res=300)
family_umap + theme_pub()
dev.off()
```

This plot shows that CsrA-containing plasmids are diverse, and are often more closely-related to non-CsrA-containing plasmids than other CsrA-containing plasmids, consistent with a model in which there is a prolonged association between CsrA and plasmids, and/or that plasmids are independently/convergently acquiring CsrA.

---

**[Back to index.](PLASMAN_index.md)**